File: src/main/java/io/prisoner/learn/thread/product/Producer.java
//package io.prisoner.learn.thread.product;
//
///**
// * DATE: 2017/8/17 23:23 <br>
// * MAIL: lbw@terminus.io <br>
// * AUTHOR: macbook
// */
//public class Producer{
//    private Integer product;
//
//    public static final Integer MAX_PRODUCT = 20;
//
//    public Integer getProduct() {
//        return product;
//    }
//
//    public void produce() {
//        if (this.product >= MAX_PRODUCT) {
//            System.out.println("producer has produced full product");
//            try {
//                wait();
//            } catch (InterruptedException e) {
//                e.printStackTrace();
//            }
//            return;
//        }
//        this.product ++;
//        System.out.println("producer is producing " + this.product + "product");
//        notifyAll();
//    }
//}

File: src/main/java/io/prisoner/learn/cookbook/chapter3/section8/Producer.kt
package io.prisoner.learn.cookbook.chapter3.section8

import java.util.concurrent.Exchanger

/**
 * DATE: 2017/8/22 00:05 <br>
 * MAIL: lbw@terminus.io <br>
 * AUTHOR: macbook
 */
class Producer(private val exchanger: Exchanger<List<String>>, private var buffer: MutableList<String>): Runnable {
    override fun run() {
        var cycle = 1
        for (i in 0 until 10) {
            println("Producer: Cycle $cycle")
            for (j in 0 until 10) {
                val message = "Event " + i *10 + j
                println("Producer: $message")
                buffer.add(message)
            }
            try {
                val buffer = exchanger.exchange(buffer)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            println("Producer: ${buffer.size}")
            cycle++
        }

    }
}